torch/nn/parallel/distributed.py

```diff
@@ -222,6 +240,66 @@ def _register_grad_hooks(self):
                     grad_acc.register_hook(self._make_param_hook(p, device_idx))
                     self._grad_accs.append(grad_acc)
 
+    def _register_nccl_grad_hook(self):
+        """
+        This function registers the callback all-reduction function for the
+        NCCL backend. All gradients will be all reduced in one single step.
+        The NCCL reduction will directly be enqueued into the
+        default CUDA stream. Therefore, no synchronization is needed.
+        """
+        # creating a new group
+        self.nccl_reduction_group_id = dist.new_group()
+
+        def reduction_fn_nccl():
+            # This function only needs to be called once
+            if not self.need_reduction:
+                return
+            self.need_reduction = False
+            all_grads = [[] for _ in range(len(self._module_copies))]
+            all_grads_coalesced = []
+
+            # Coalesce all the gradients
+            # TODO: Add mixed precision support here
+            for idx, module in enumerate(self._module_copies):
+                for param in module.parameters():
+                    if not param.requires_grad or param.grad is None:
+                        continue
+                    if param.grad.requires_grad:
+                        raise RuntimeError("DistributedDataParallel only works "
+                                           "with gradients that don't require "
+                                           "grad")
+                    # Adding the gradients for reduction
+                    all_grads[idx].append(param.grad.data)
+                with torch.cuda.device(self.device_ids[idx]):
+                    dev_grads_coalesced = _flatten_dense_tensors(all_grads[idx])
+                    all_grads_coalesced.append(dev_grads_coalesced)
+
+            # This single op will do all-reduce on all GPUs utilizing multiple
+            # all reduce rings when we have more than one fast IB interfaces.
+            # We will only use device 0's results, but this single op should be
+            # faster than doing the following two operation sequentially:
+            # (1) intra-node reduce to lead GPU, followed by
+            # (2) inter-node allreduce for all the first lead GPUs in all nodes
+            dist.all_reduce_multigpu(all_grads_coalesced,
+                                     group=self.nccl_reduction_group_id)
+
+            # Now only work on the first lead GPU
+            all_grads_coalesced[0] /= dist.get_world_size()
+            for grad, reduced in \
+                    zip(all_grads[0],
+                        _unflatten_dense_tensors(all_grads_coalesced[0],
+                                                 all_grads[0])):
```

Comment:
ouch. Let's do something about this line, because it's impossible to read. How about this:
```python
iter = zip(...)
for grad, reduced in iter:
    ...
```